util/registrationdb/registrationdb.go

```diff
@@ -0,0 +1,356 @@
+// Package registrationdb provides storage and lookup mechanisms for NSQ topic producers.
+// The RegistrationDB data structure provides a mapping from a topic and channel pair
+// to a registry of producers.
+package registrationdb
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// RegistrationDB provides storage and lookup mechanisms for NSQ topic producers.
+type RegistrationDB struct {
+	mtx     sync.RWMutex
+	data    map[RegistryKey]Registry
+	clients map[string]*Producer
```

Comment:
I'm not a big fan of building explicit knowledge of clients into `RegistrationDB`.  It was intentionally designed to be agnostic to the data that we put in, for the most part.

Granted, it is only used inside `nsqlookupd` (and now `nsqd`) but I think that approach is still relevant.

Basically, I don't see the benefit of conflating these sets of changes in this PR.  Ideally the semantics and API of `RegistrationDB` would remain the same (save for obvious cleanup/bugs/etc.) and this PR would focus on the actual core changes to introduce gossip.  I think this will reduce the noise and make it easier to actually land this PR confidently.

What do you think?